src/lib/message/ExchangeContext.cpp

```diff
@@ -0,0 +1,1664 @@
+        (profileId == chip::Profiles::kChipProfile_Common && msgType == chip::Profiles::Common::kMsgType_Null))
+    {
+        if (kChipMessageVersion_Unspecified == mMsgProtocolVersion)
+        {
+            mMsgProtocolVersion = msgInfo->MessageVersion = kChipMessageVersion_V2;
+        }
+        else
+        {
+            VerifyOrExit(mMsgProtocolVersion == kChipMessageVersion_V2, err = CHIP_ERROR_WRONG_MSG_VERSION_FOR_EXCHANGE);
+        }
+    }
+#endif
+    if (kChipMessageVersion_Unspecified == mMsgProtocolVersion)
+    {
+        mMsgProtocolVersion = msgInfo->MessageVersion = kChipMessageVersion_V1;
+    }
+    else
+    {
+        msgInfo->MessageVersion = mMsgProtocolVersion;
+    }
+
+    // Prevent sending if the context was associated with a connection that is now closed.
+    VerifyOrExit(!IsConnectionClosed(), err = CHIP_ERROR_NOT_CONNECTED);
+
+    // TODO: implement support for retransmissions.
+
+    // flag validation
+    if (sendFlags & kSendFlag_RetransmissionTrickle)
+    {
+        // We do not allow RMP to be used when Trickle retransmission is requested
+        if (sendFlags & kSendFlag_RequestAck)
+        {
+            ExitNow(err = CHIP_ERROR_INVALID_ARGUMENT);
+        }
+
+        // We do not support trickle retrasnmissions over
+        // connection-oriented exchanges
+        VerifyOrExit(Con == NULL, err = CHIP_ERROR_INVALID_ARGUMENT);
+
+        if (0 == RetransInterval)
+        { // we're not retransmitting, do not hold onto the buffer
+            sendFlags &= ~kSendFlag_RetainBuffer;
+        }
+        else
+        {
+            sendFlags |= kSendFlag_RetainBuffer;
+            msg = msgBuf;
+        }
+    }
+
+    // Add the exchange header to the message buffer.
+    ChipExchangeHeader exchangeHeader;
+    memset(&exchangeHeader, 0, sizeof(exchangeHeader));
+
+    err = EncodeExchHeader(&exchangeHeader, profileId, msgType, msgBuf, sendFlags);
+    SuccessOrExit(err);
+
+    // If a response message is expected...
+    if ((sendFlags & kSendFlag_ExpectResponse) != 0)
+    {
+        // Only one 'response expected' message can be outstanding at a time.
+        VerifyOrExit(!IsResponseExpected(), err = CHIP_ERROR_INCORRECT_STATE);
+
+        SetResponseExpected(true);
+
+        // Arm the response timer if a timeout has been specified.
+        if (ResponseTimeout > 0)
+        {
+            err = StartResponseTimer();
+            SuccessOrExit(err);
+        }
+    }
+
+    // Fill in appropriate message header flags
+    if (sendFlags & kSendFlag_DelaySend)
+        msgInfo->Flags |= kChipMessageFlag_DelaySend;
+    // FIXME: RS: possibly unnecessary, should addref instead
+    if (sendFlags & kSendFlag_RetainBuffer)
+        msgInfo->Flags |= kChipMessageFlag_RetainBuffer;
+    if (sendFlags & kSendFlag_AlreadyEncoded)
+        msgInfo->Flags |= kChipMessageFlag_MessageEncoded;
+    if (sendFlags & kSendFlag_ReuseMessageId)
+        msgInfo->Flags |= kChipMessageFlag_ReuseMessageId;
```

Comment:
I thought secure channel didn't want message ID reuse? @RyanTheOptimist @balducci-apple

Reply:
This is a straight copy from weave, no attempts at updating. I think we should update in separate PRs (lots of cleanup needed in general).

I expect the entire security model needs to be plugged out and replaced with SCTT.

Reply:
@woody-apple 

This is just the strait copy of the missing parts in WML, these files are not the final version and not get linked at this moment.

We are working on integrating the existing session management logic to this baseline in separate PRs. Here is the open items list for this effort:
https://github.com/project-chip/connectedhomeip/labels/cleanup_wml_baseline

Reply:
Yes, this should be dropped.  This is a bit of a vestigial code that we're not pulling into CHIP (this was aimed at a self regulating group communication pattern that implemented a trickle algorithm, and we're happy to pull that out).  To that end, we should pull out: 

- The flag as identified here
- `ResendMessage` implementation (in ExchangeContext only)
- any if statements guarded with references to Trickle
- TeardownTrickleRetransmit
- StartTimerT
- StartTimerTau
- TimerT
- TimerTau
- HandleTrickleMessage
- SetupTrickleRetransmit